<div style="font-family: 'Arial', sans-serif; font-size: 48px; font-weight: bold; text-align: center; color: #ffcc00; background-color: #2a2a2a; padding: 30px; border: 3px solid #444; border-radius: 15px; box-shadow: 0 4px 10px rgba(0, 0, 0, 0.3); margin-bottom: 20px;">
    Data Mining
</div>

<div style="font-family: 'Arial', sans-serif; font-size: 48px; font-weight: bold; text-align: center; color: #ff9900; background-color: #333333; padding: 20px; border: 3px solid #444; border-radius: 15px; box-shadow: 0 4px 10px rgba(0, 0, 0, 0.3); margin-bottom: 20px;">
    Lab 04
</div>

<div style="font-family: 'Arial', sans-serif; font-size: 48px; font-weight: bold; text-align: center; color: #ff6600; background-color: #1c1c1c; padding: 20px; border: 3px solid #444; border-radius: 15px; box-shadow: 0 4px 10px rgba(0, 0, 0, 0.3); margin-bottom: 20px;">
    Syed Mansoor ul Hassan Bukhari
</div>

<div style="font-family: 'Arial', sans-serif; font-size: 48px; font-weight: bold; text-align: center; color: #ffcc00; background-color: #222222; padding: 20px; border: 3px solid #444; border-radius: 15px; box-shadow: 0 4px 10px rgba(0, 0, 0, 0.3); margin-bottom: 20px;">
    BSAI - VII
</div>

<div style="font-family: 'Arial', sans-serif; font-size: 48px; font-weight: bold; text-align: center; color: #ff3300; background-color: #1c1c1c; padding: 20px; border: 3px solid #444; border-radius: 15px; box-shadow: 0 4px 10px rgba(0, 0, 0, 0.3);">
    2021-25
</div>

<div style="font-family: 'Arial', sans-serif; font-size: 48px; font-weight: bold; text-align: center; color: #3399ff; background-color: #1c1c1c; padding: 20px; border: 3px solid #444; border-radius: 15px; box-shadow: 0 4px 10px rgba(0, 0, 0, 0.3);">
    <span style="position: relative; top: -10px;">12<sup>th</sup></span> Nov, 2024
</div>


<div style="font-family: 'Arial', sans-serif; font-size: 20px; text-align: left; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
    <strong style="color: #ffcc00;">Question 01: </strong>
    <span style="color: #cccccc;">Find most similar documents from the following data matrix using Cosine Similarity measure.</span>
</div>
<img src="Pictures/Picture2.jpg" />

<div style="font-family: 'Arial', sans-serif; font-size: 20px; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
    <strong style="color: #ffcc00;">Finding Most Similar Documents Using Cosine Similarity</strong><br><br>
    <span style="color: #cccccc;">In this section, we calculate the similarity between documents based on their numerical features using the <span style="color: #ffcc00;">Cosine Similarity</span> measure. Cosine similarity measures the cosine of the angle between two vectors, providing a way to quantify the degree of similarity between them.</span></br>
    <strong style="color: #ffcc00;">1. Prepare the Data</strong></br>
    <span style="color: #cccccc;">Each document is represented by numerical features (columns), which are treated as vectors in a multi-dimensional space. For example, the "Team", "Coach", "Hockey", "Baseball", "Soccer", and other columns represent different dimensions of the document's feature vector.</span></br>
    <strong style="color: #ffcc00;">2. Calculate Cosine Similarity</strong></br>
    <span style="color: #cccccc;">The cosine similarity between two vectors <span style="color: #ffcc00;">A</span> and <span style="color: #ffcc00;">B</span> is given by the formula:</span></br>
    <pre style="color: #cccccc; background-color: #333333; padding: 10px;">
        cosine_similarity(A, B) = (A ⋅ B) / (||A|| ||B||)
    </pre>
    <span style="color: #cccccc;">Where:</span></br>
    <ul style="color: #cccccc;">
        <li><span style="color: #ffcc00;">A ⋅ B</span> is the dot product of vectors <span style="color: #ffcc00;">A</span> and <span style="color: #ffcc00;">B</span>,</li>
        <li><span style="color: #ffcc00;">||A||</span> and <span style="color: #ffcc00;">||B||</span> are the magnitudes (norms) of the vectors <span style="color: #ffcc00;">A</span> and <span style="color: #ffcc00;">B</span>.</li>
    </ul>
    <span style="color: #cccccc;">The cosine similarity score ranges from -1 (completely dissimilar) to 1 (identical vectors), with 0 indicating orthogonal (no similarity) vectors.</span></br>
        <strong style="color: #ffcc00;">3. Compute Pairwise Cosine Similarity</strong><br><span style="color: #cccccc;">To find the most similar documents, we calculate the cosine similarity between every pair of documents in the dataset. The result is a similarity matrix that shows the pairwise similarity scores.</span>
</div>


<div style="font-family: 'Arial', sans-serif; font-size: 20px; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
<center>
  <span style="color:#4682B4; font-size:22px; font-weight: bold; padding: 5px;">Import Libraries</span>
</center>
</div>

In [1]:
import pandas as pd
from sklearn.metrics.pairwise import cosine_similarity

<div style="font-family: 'Arial', sans-serif; font-size: 20px; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
<center>
  <span style="color:#4682B4; font-size:22px; font-weight: bold; padding: 5px;">Create DataFrame</span>
</center>
</div>

In [2]:
df = pd.DataFrame({
    "Document": ['Document1', 'Document2', 'Document3', 'Document4'],
    'Team': [5, 3, 0, 0],
    'Coach': [0, 0, 7, 1],
    'Hockey': [3, 2, 0, 0],
    'Baseball': [0, 0, 2, 0],
    'Soccer': [2, 1, 1, 1],
    'Penalty': [0, 1, 0, 2],
    'Score': [0, 0, 0, 2],
    'Win': [2, 1, 3, 0],
    'Loss': [0, 0, 0, 3],
    'Season': [0, 1, 0, 0]
})

<div style="font-family: 'Arial', sans-serif; font-size: 20px; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
<center>
  <span style="color:#4682B4; font-size:22px; font-weight: bold; padding: 5px;">Drop the 'Document' column as it's not part of the numerical matrix</span>
</center>
</div>

In [3]:
matrix = df.drop(columns=['Document'])

In [4]:
df

Document  Team  Coach  Hockey  Baseball  Soccer  Penalty  Score  Win  \
0  Document1     5      0       3         0       2        0      0    2   
1  Document2     3      0       2         0       1        1      0    1   
2  Document3     0      7       0         2       1        0      0    3   
3  Document4     0      1       0         0       1        2      2    0   

   Loss  Season  
0     0       0  
1     0       1  
2     0       0  
3     3       0

In [5]:
matrix

Team  Coach  Hockey  Baseball  Soccer  Penalty  Score  Win  Loss  Season
0     5      0       3         0       2        0      0    2     0       0
1     3      0       2         0       1        1      0    1     0       1
2     0      7       0         2       1        0      0    3     0       0
3     0      1       0         0       1        2      2    0     3       0

<div style="font-family: 'Arial', sans-serif; font-size: 20px; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
<center>
  <span style="color:#4682B4; font-size:22px; font-weight: bold; padding: 5px;">Compute cosine similarity</span>
</center>
</div>

In [6]:
cos_sim = cosine_similarity(matrix)

In [7]:
cos_sim

array([[1.        , 0.93560149, 0.15552316, 0.07079923],
       [0.93560149, 1.        , 0.12222647, 0.16692447],
       [0.15552316, 0.12222647, 1.        , 0.23122933],
       [0.07079923, 0.16692447, 0.23122933, 1.        ]])

<div style="font-family: 'Arial', sans-serif; font-size: 20px; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
<center>
  <span style="color:#4682B4; font-size:22px; font-weight: bold; padding: 5px;">Create a DataFrame of cosine similarities for better visualization</span>
</center>
</div>

In [8]:
cos_sim_df = pd.DataFrame(cos_sim, index=df['Document'], columns=df['Document'])

In [9]:
cos_sim_df

Document   Document1  Document2  Document3  Document4
Document                                             
Document1   1.000000   0.935601   0.155523   0.070799
Document2   0.935601   1.000000   0.122226   0.166924
Document3   0.155523   0.122226   1.000000   0.231229
Document4   0.070799   0.166924   0.231229   1.000000

<div style="font-family: 'Arial', sans-serif; font-size: 20px; color: #ffffff; background-color: #2d2d2d; padding: 25px; border: 3px solid #555555; border-radius: 15px; box-shadow: 0 6px 15px rgba(0, 0, 0, 0.4); margin-bottom: 25px;">
    <strong style="color: #ffcc00;">Cosine Similarity Results</strong><br><br>
    <span style="color: #cccccc;">Here are the cosine similarity values between the documents:</span><br><br>
    <strong style="color: #ffcc00;">Document1</strong>:<br>
    <span style="color: #cccccc;">- Similarity with Document2: 0.9356 (High)</span><br>
    <span style="color: #cccccc;">- Similarity with Document3: 0.1555 (Low)</span><br>
    <span style="color: #cccccc;">- Similarity with Document4: 0.0708 (Very Low)</span><br><br>
    <strong style="color: #ffcc00;">Document2</strong>:<br>
    <span style="color: #cccccc;">- Similarity with Document1: 0.9356 (High)</span><br>
    <span style="color: #cccccc;">- Similarity with Document3: 0.1222 (Very Low)</span><br>
    <span style="color: #cccccc;">- Similarity with Document4: 0.1669 (Low)</span><br><br>
    <strong style="color: #ffcc00;">Document3</strong>:<br>
    <span style="color: #cccccc;">- Similarity with Document1: 0.1555 (Low)</span><br>
    <span style="color: #cccccc;">- Similarity with Document2: 0.1222 (Very Low)</span><br>
    <span style="color: #cccccc;">- Similarity with Document4: 0.2312 (Low)</span><br><br>
    <strong style="color: #ffcc00;">Document4</strong>:<br>
    <span style="color: #cccccc;">- Similarity with Document1: 0.0708 (Very Low)</span><br>
    <span style="color: #cccccc;">- Similarity with Document2: 0.1669 (Low)</span><br>
    <span style="color: #cccccc;">- Similarity with Document3: 0.2312 (Low)</span><br><br>
</div>
